### ISSUE TITLE
Dynamic fake latency

### ISSUE THREAD

**Issue by SylveonBottle** [nullworks/cathook#1087]
As it currently stands, if you enable backtrack, you can see all of the backtrack chams behind the enemy. But if you're chasing someone and you're too far behind them, you won't be able to backstab/backtrack them because they're too far ahead. But if you increase your fake latency, the backtrack would extend and you could easily hit them lol.

It would be sick if cathook could automatically increase your fake latency so you can backstab enemies you are chasing.

Method of doing this-
Keep track of player positions for the entire 1000ms. Leave backtracking at 0. 
If you are behind an enemy (their 1000ms-th position is closer than their current backtrack), switch fake latency up to 1000. 
If setting it to 1000ms would result in backtracking chams to be BEHIND you, set it to 800ms instead. So on and so forth. Dynamic. Just change fake latency to whichever previous position is closer to you.

>You're hiding behind enemy lines.
>A medic comes through, running to the front lines.
>You chase after him. But he's just slightly too fast.
>You have backtrack enable. But he's about 10 feet infront of you, so you can't hit the backtrack.
>Cathook sees that the medic's 700ms backtrack is closer to you (1000ms is behind you, 600ms is too far ahead) so automatically enables fake latency to 700.
>You're able to backtrack him, so cathook's autobackstab stabs him.
>Cathook sees that he's dead, and disables fake latency.

**Comment by BenCat07** [comment]
Changing latency causes you to get rubberbanded so that is not really an option

**Comment by Footsiefat** [comment]
You can normally chance it so set a limit of how often it can change it?

**Comment by SylveonBottle** [comment]
Is this suggestion impossible?

**Comment by TotallyNotElite** [comment]
No, reopening, this is rather useless IMO but a feature suggestion is a feature suggestion

**Comment by BenCat07** [comment]
It is possible, however not fully

**Comment by BenCat07** [comment]
I could make t so it adjusts the ping based on your situation, but the rampup will still exist

**Comment by iraizo** [comment]
Im not even sure how dynamic fake latency is adding something to cathook, what advantage does that give you?

**Comment by TotallyNotElite** [comment]
Longer melee range on demand

**Comment by Footsiefat** [comment]
Yep. The thing is sometimes the fake latency will give you too much range making it hard to hit a target

**Comment by iraizo** [comment]
> Longer melee range on demand

Couldn't you just toggle fake latency if you want to melee somebody or do i not get what "dynamic fatency" is supposed to be?

**Comment by TotallyNotElite** [comment]
Some people are lazy.

**Comment by SylveonBottle** [comment]
> 
> 
> I could make it so it adjusts the ping based on your situation, but the rampup will still exist

>record full 1000ms of user recordings
>put fakelag to 200
>if meelee is out, and if the user is inside of the 400-1000ms fakelag, increase fakelag to 600/800ms to allow melee to hit
>change back once we're outside of the 400-1000ms recorded positions



> Im not even sure how dynamic fake latency is adding something to cathook, what advantage does that give you?

>using spy
>using backtrack
>using 200 fake latency
>chasing medic
>am behind medic
>can't reach medic
>still chasing medic
>"if I quickly open up the menu and switch my fakelatency to 400ms I could've hit this medic, but by stopping and doing that the medic would have gotten even further away"




> Couldn't you just toggle fake latency if you want to melee somebody or do i not get what "dynamic fatency" is supposed to be?

If you already have fake latency to 200ms, you can't just "toggle" it off to put it at 800ms. I guess you could use akeybind but at that point most people would just not trouble with it / forget such a keybind exists. If it was automated, that'd be amazing. Like automating crits against raged players. It's useful because sometimes people just don't think about it fast enough. Plus automating this would allow it to switch back without people forgetting "wait what's my fake latency at" and spam the button a few times checking their ping before giving up and going back to the menu fuckin' anyways to see if it's set to 200 or 800 now.

**Comment by iraizo** [comment]
I guess, my approach would be adding an automatic toggle for fake latency everytime you have a melee in your hand but then again im lazy.

**Comment by iraizo** [comment]
> If you already have fake latency to 200ms, you can't just "toggle" it off to put it at 800ms. I guess you could use akeybind but at that point most people would just not trouble with it / forget such a keybind exists. If it was automated, that'd be amazing. Like automating crits against raged players. It's useful because sometimes people just don't think about it fast enough. Plus automating this would allow it to switch back without people forgetting "wait what's my fake latency at" and spam the button a few times checking their ping before giving up and going back to the menu fuckin' anyways to see if it's set to 200 or 800 now.

Ah now i see your point, excuse me for closing this issue then, it is a good suggestion,

**Comment by SylveonBottle** [comment]
> 
> 
> I guess, my approach would be adding an automatic toggle for fake latency everytime you have a melee in your hand but then again im lazy.

Isn't this already a thing? Set backtrack to melee only and you're good? The issue is sometimes you'll have 800ms fake latency and be TOO CLOSE to the enemy to hit them, in this situation it would've been good to lower fake latency.. which you can't do unless you use specific keybinds or automate this.

**Comment by SylveonBottle** [comment]
The whole point of automation is seeing if we can do it / doing it for us. If you could "just bind `cat_toggle fakelatency 200 800` and do it manually", you could "just bind wasd to move, and move manually", or "Just bind aim to mouse, and hit heads manually". Literally why even code an aimbot if you can just click on heads? _Because automation is cool as shit, c'mon._

We don't need an autozoom / autounzoom. But somebody thought it was cool to implement, so it was implemented.  Is it my fault for thinking that implimenting something that automates switching fake latency values in order to hit a player further / closer than you could've is awesome? Is it stupid / hard to make / no one would really use it?

**Comment by iraizo** [comment]
Stop getting so defensive about it, everybody agrees that this suggestion is fine, i just misinterpreted the issue

**Comment by BenCat07** [comment]
I will add a "smart latency" option with the new backtrack i guess where it picks the best value bur prefers your set latency

**Comment by SylveonBottle** [comment]
> I will add a "smart latency" option with the new backtrack i guess where it picks the best value bur prefers your set latency
What do you think would be a good method of deciding the latency?



> Stop getting so defensive about it, everybody agrees that this suggestion is fine, i just misinterpreted the issue
I have no life and too much time on my hands, tbh this thread was the most exciting thing to happen to me all weak 😭

**Comment by Footsiefat** [comment]
@BenCat07 Also I should let you know while I'm here, the auto backstab that uses backtracking often ends up facestabing and I've noticed I need to often rotate my view angle to line up with their backs.

So maybe looking into those facestabs 
Or making auto backstab auto rotate to the correct angle :),

Would be awesome thanks.

**Comment by iraizo** [comment]
> @BenCat07 Also I should let you know while I'm here, the auto backstab that uses backtracking often ends up facestabing and I've noticed I need to often rotate my view angle to line up with their backs.

Related to https://github.com/nullworks/cathook/pull/1381
